server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ContinuousComputationTests.java

```diff
@@ -134,4 +136,57 @@ protected void processInput(Object input) {
         await.run();
         assertBusy(() -> assertFalse(computation.isActive()));
     }
+
+    public void testFailureHandling() {
+        final var input1 = new Object();
+        final var input2 = new Object();
+
+        final var failureCount = new AtomicInteger();
+
+        final var computation = new ContinuousComputation<>(r -> {
+            try {
+                r.run();
+            } catch (AssertionError e) {
+                assertEquals("simulated", asInstanceOf(RuntimeException.class, e.getCause()).getMessage());
+                failureCount.incrementAndGet();
+            }
+        }) {
+            @Override
+            protected void processInput(Object input) {
+                if (input == input1) {
+                    onNewInput(input2);
+                }
+                throw new RuntimeException("simulated");
+            }
+
+            @Override
+            public String toString() {
+                return "test computation";
+            }
+        };
+
+        final var mockLogAppender = new MockLogAppender();
+        mockLogAppender.addExpectation(
+            new MockLogAppender.SeenEventExpectation(
+                "error log",
+                ContinuousComputation.class.getCanonicalName(),
+                Level.ERROR,
+                "unexpected error processing [test computation]"
+            )
+        );
+
+        try (var ignored = mockLogAppender.capturing(ContinuousComputation.class)) {
+            computation.onNewInput(input1);
+            mockLogAppender.assertAllExpectationsMatched();
+        }
+
+        // check that both inputs were processed
+        assertEquals(2, failureCount.get());
```

Comment:
I was originally confused as there is only one call to `onNewInput` that results in 2 errors.
I wonder if instead we should not chain executions in `processInput` and instead have 1 input that is failing (and emit error log) and one that is succeeding?

Reply:
:+1: sure, see cded334a0b805350c6fcd65d127143ad042951ab